### PR TITLE
MH-12813: Add audio and video track selection to video editor

### PR DESCRIPTION
**Pull request by pmiddend**
This PR adds a new tab “Tracks” to the video editor that allows the user to select which audio and video tracks will be published. In this new tab, the user simply clicks on the video track (represented by a
thumbnail) or the audio track (represented by its waveform) to hide it. Sanity checks are included, so its not possible to hide _every_ video and audio track there is, for example.

40481539-0407fbda-5f52-11e8-872c-704e588a36d0

The information about which tracks are hidden is stored in asset manager properties, which are then injected into the running workflow(s). Thus, this PR depends on [193](#193) (and is rebased on it).

This _theoretically_ works for an arbitrary number of tracks in the media package, but it currently is constrained to one and two tracks (the most common use case). The allowed flavors flavors are configurable.

To make this possible, a specialized workflow operation handler `SelectTracksWorkflowOperationHandler` had to be written, which is capable of removing audio tracks from multiple input tracks, as well as muxing a video track and an audio track from two different input
tracks (so it’s very similar to the `PrepareAV` WOH).

This work was sponsored by SWITCH.

### CODE REVIEW

**Comment by JulianKniephoff** [comment]
I am looking at this in context of another possibly related issue I am working on for the ETH and I noticed the following:

image

This happened when selecting only one video stream, but two audio streams, obviously; the problem is that this message does not disappear when I then do "the right thing" and successfully publish the event.

Also I find it kind of confusing that this message appears only after I hit the "do the thing button", but the error message for deselecting for example all audio streams appears instantly (and in a different position).

**Comment by pmiddend** [comment]
@JulianKniephoff You're absolutely right. I fixed the issues you mentioned.

**Comment by staubesv** [comment]
Hi @pmiddend Please note that this PR currently has merge conflicts. Could you please resolve them?

**Comment by pmiddend** [comment]
@staubesv Thanks for noticing. Fixed.

**Comment by staubesv** [comment]
Hi @pmiddend, please note that this PR has merge conflicts again.

**Comment by krinnewitz** [comment]
Hi @staubesv, Conflicts should now be fixed (again).

**Comment by staubesv** [comment]
Thanks ;-)

File: etc/encoding/opencast-movies.properties [hunk 1]
```diff
@@ -48,7 +48,7 @@ profile.mux-av.work.name = mux audio and video
 profile.mux-av.work.input = stream
 profile.mux-av.work.output = visual
 profile.mux-av.work.suffix = -work.#{in.video.suffix}
-profile.mux-av.work.ffmpeg.command = -i #{in.video.path} -i #{in.audio.path} -shortest -c copy #{out.dir}/#{out.name}#{out.suffix}
+profile.mux-av.work.ffmpeg.command = -i #{in.video.path} -i #{in.audio.path} -shortest -c copy -map 0:v:0 -map 1:a:0 #{out.dir}/#{out.name}#{out.suffix}
```
Comment (staubesv) [inline etc/encoding/opencast-movies.properties:51]: As this encoding profile is hard-coded in WOH prepare-av, you might want to introduce another encoding profile for WOH tracks-select. Or is this change supposed to be useful for WOH prepare-av, too?

Comment (pmiddend) [reply etc/encoding/opencast-movies.properties:51]: Ah yes. So when the prepare-av workflow receives a mediapackage with two tracks, and it's supposed to mux these together (so take the audio subtrack of track A and the video subtrack of track B), it actually strips the video subtrack from A and then merges the two streams together. This is unnecessary, since ffmpeg can mux the two streams with a single command. I asked why this isn't being done in the profile above, and Greg replied that it might be because the code is ancient (meaning from times when ffmpeg wasn't able to mux as intelligently as it can, now).
To be extremely sure, I could introduce another profile.

Comment (staubesv) [reply etc/encoding/opencast-movies.properties:51]: Thanks for the info! In that case, just modifying the existing encoding profile should be fine.

**Comment by staubesv** [comment]
Hi @pmiddend I've just merged #264 which caused merged conflicts in this PR. Could you please resolve them?

**Comment by krinnewitz** [comment]
Hi @staubesv The merge conflict is resolved. Travis build seems to fail due to random DNS lookup error.

File: modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/SelectTracksWorkflowOperationHandler.java [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+  private enum AudioMuxing {
+    NONE("none"), FORCE("force"), DUPLICATE("duplicate");
+
+    private final String value;
+
+    AudioMuxing(final String value) {
+      this.value = value;
+    }
+
+    static AudioMuxing fromConfigurationString(final String s) {
```
Comment (lkiesow) [inline modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/SelectTracksWorkflowOperationHandler.java]: Isn't this basically what the `enum` does anyway and you could e.g. just use `AudioMuxing.valueOf(…)` instead of this custom code?

File: docs/guides/admin/docs/workflowoperationhandlers/select-tracks-woh.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+target-tags       | download  | The tags applied to all target tracks
+audio-muxing      | force     | Move single-audio media packages to a specific track (see below)
+force-target      | presenter     | Target track for the `force` setting for `audio-muxing`
+
+\* mandatory configuration key
+
+Audio Muxing
+-----------------
+
+The optional `audio-muxing` parameter has three possible values: `none` (same as omitting the option), `force` and
```
Comment (lkiesow) [inline docs/guides/admin/docs/workflowoperationhandlers/select-tracks-woh.md]: There is no description for `none`? What happens if `none` is specified (or the configuration is omitted)?

File: docs/guides/admin/docs/workflowoperationhandlers/select-tracks-woh.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+`duplicate`.
+
+### `force` ###
+
+The parameter value `force` only applies to media packages that have exactly one non-hidden audio track. For media packages
```
Comment (lkiesow) [inline docs/guides/admin/docs/workflowoperationhandlers/select-tracks-woh.md]: The usage of tracks in this section is highly confusing. You are confusing tracks with streams.

For example:

> …a media package with two tracks, “presenter” and “presentation”, and the audio track of “presenter” is hidden.

actually means

> …a media package with two tracks, “presenter” and “presentation”, and the audio *stream* of “presenter” is hidden.

because if you hide a track's track, you hide itself, don't you? Note that, of course, it's possible to have two tracks flavored the same (e.g. two presenter tracks) but I don't think you mean that.

File: modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/SelectTracksWorkflowOperationHandler.java [hunk 5]
```diff
@@ -0,0 +1,500 @@
...
+
+    // First case: We have only tracks with non-hidden video streams. So we keep them all and possibly cut away audio.
+    if (allNonHidden(augmentedTracks, SubTrack.VIDEO)) {
+      final Optional<AudioMuxing> audioMuxing = getConfiguration(workflowInstance, CONFIG_AUDIO_MUXING)
+              .map(AudioMuxing::fromConfigurationString);
+      // For both special options, we need to find out if we have exactly one audio track present
+      final Optional<AugmentedTrack> singleAudioTrackOpt = findSingleAudioTrack(augmentedTracks);
+      final boolean multipleVideo = augmentedTracks.size() > 1;
+      if (multipleVideo && audioMuxing.map(m -> m == AudioMuxing.DUPLICATE).orElse(Boolean.FALSE) && singleAudioTrackOpt
+              .isPresent()) {
```
Comment (lkiesow) [inline modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/SelectTracksWorkflowOperationHandler.java]: Maybe use
```java
AudioMuxing.DUPLICATE.equals(audioMuxing.orElse(null))
```

instead of

```java
audioMuxing.map(m -> m == AudioMuxing.DUPLICATE).orElse(Boolean.FALSE)
```

to be less obscure or, even better, not use an Optional at all since the default value is basically `AudioMuxing.EMPTY` anyway, making this not really optional.

File: modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/SelectTracksWorkflowOperationHandler.java [hunk 6]
```diff
@@ -0,0 +1,500 @@
...
+    } else {
+      // Second case: we have exactly one video track that is not hidden (hopefully, because for all other cases there
+      // were no requirements given).
+      final MuxResult muxResult = muxSingleVideoTrack(mediaPackage, augmentedTracks);
+      result.add(muxResult);
+    }
+
+    final MediaPackageElementFlavor targetTrackFlavor = MediaPackageElementFlavor.parseFlavor(StringUtils.trimToNull(
+            getConfiguration(workflowInstance, "target-flavor")
+                    .orElseThrow(() -> new IllegalStateException("Target flavor not specified"))));
```
Comment (lkiesow) [inline modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/SelectTracksWorkflowOperationHandler.java]: Fail fast. No need to handle all the media file operations if we finally fail with a configuration error anyway.

File: modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/SelectTracksWorkflowOperationHandler.java [hunk 7]
```diff
@@ -0,0 +1,500 @@
...
+
+        if (singleAudioTrack.track != forceTargetTrack.track) {
+          // Copy it over...
+          final TrackJobResult muxResult = mux(forceTargetTrack.track, singleAudioTrack.track, mediaPackage);
+          result.add(muxResult);
+
+          // ...and remove the original
+          final TrackJobResult hideAudioResult = hideAudio(singleAudioTrack.track, mediaPackage);
+          result.add(hideAudioResult);
+        }
```
Comment (lkiesow) [inline modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/SelectTracksWorkflowOperationHandler.java]: You need to copy tracks you do not mux as well. Otherwise, the output of this operation is impossible to predict.

Example:

- You have two video files `presenter/source` and `presentation/source`. The first has video only, the second has video and audio.
- You set this operation to force audio in `presentation` and set the target flavor to `*/selected`
- The operation will run, claim success but since the audio was already in the right place, there is no track with the flavor `*/selected`
- Result: The next operation will likely fail.

here is a simple patch which fixes the problem. I'm not sure if I covered all edge cases:

```diff
diff --git a/modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/SelectTracksWorkflowOperationHandler.java b/modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/SelectTracksWorkflowOperationHandler.java
index dd135ddbd..385687bdd 100644
--- a/modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/SelectTracksWorkflowOperationHandler.java
+++ b/modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/SelectTracksWorkflowOperationHandler.java
@@ -39,11 +39,13 @@ import org.opencastproject.workflow.api.WorkflowOperationResult;
 import org.opencastproject.workflow.api.WorkflowOperationTagUtil;
 import org.opencastproject.workspace.api.Workspace;
 
+import org.apache.commons.io.FilenameUtils;
 import org.apache.commons.lang3.StringUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.io.IOException;
+import java.net.URI;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
@@ -51,6 +53,7 @@ import java.util.List;
 import java.util.Optional;
 import java.util.SortedMap;
 import java.util.TreeMap;
+import java.util.UUID;
 import java.util.function.Consumer;
 import java.util.stream.Collectors;
 
@@ -273,6 +276,8 @@ public class SelectTracksWorkflowOperationHandler extends AbstractWorkflowOperat
           if (t.track != singleAudioTrack.track) {
             final TrackJobResult jobResult = mux(t.track, singleAudioTrack.track, mediaPackage);
             result.add(jobResult);
+          } else {
+            result.add(copyTrack(t.track));
           }
         }
       } else if (multipleVideo && audioMuxing.map(m -> m == AudioMuxing.FORCE).orElse(Boolean.FALSE)
@@ -302,6 +307,15 @@ public class SelectTracksWorkflowOperationHandler extends AbstractWorkflowOperat
           // ...and remove the original
           final TrackJobResult hideAudioResult = hideAudio(singleAudioTrack.track, mediaPackage);
           result.add(hideAudioResult);
+        } else {
+          result.add(copyTrack(singleAudioTrack.track));
+        }
+
+        // Just copy the rest of the tracks to ensure they got the correct output flavor
+        for (final AugmentedTrack augmentedTrack : augmentedTracks) {
+          if (augmentedTrack.track != singleAudioTrack.track && augmentedTrack.track != forceTargetTrack.track) {
+            result.add(copyTrack(augmentedTrack.track));
+          }
         }
       } else {
         // No special options selected, or conditions for special options don't match.
@@ -497,4 +511,27 @@ public class SelectTracksWorkflowOperationHandler extends AbstractWorkflowOperat
       return new AugmentedTrack(t, hideAudio, hideVideo);
     }).collect(Collectors.toList());
   }
+
+  private TrackJobResult copyTrack(Track track) throws WorkflowOperationException {
+    Track copiedTrack = (Track) track.clone();
+    copiedTrack.setIdentifier(UUID.randomUUID().toString());
+    try {
+      // Generate a new filename
+      String targetFilename = copiedTrack.getIdentifier();
+      String extension = FilenameUtils.getExtension(track.getURI().getPath());
+      if (extension.length() > 0) {
+        targetFilename += "." + extension;
+      }
+
+      // Copy the files on dis and put them into the working file repository
+      logger.debug("Start copying element {}.", track.getURI());
+      URI newUri = workspace.put(track.getMediaPackage().getIdentifier().toString(), copiedTrack.getIdentifier(),
+              targetFilename, workspace.read(track.getURI()));
+      copiedTrack.setURI(newUri);
+    } catch (IOException | NotFoundException e) {
+      throw new WorkflowOperationException(String.format("Error while copying track %s", track.getIdentifier()), e);
+    }
+
+    return new TrackJobResult(copiedTrack, 0);
+  }
 }
```

**Review by lkiesow — CHANGES_REQUESTED** [review]
This is mainly a review of the select-tracks woh which has some broken edge cases. Note that the operations name is highly confusing since it does not select tracks but video and audio streams.

**Comment by pmiddend** [comment]
Thanks @lkiesow for the excellent review. I've pushed the following changes:

  - I fixed the markdown docs, after banging my head for not realizing "stream" was the term I was looking for all along. I also added a section about the default behavior of `none` for `audio-muxing`. 
  - I applied your patch containing `copyTrack` and tested it. Looked good so far, and I think you got all the edge cases.
  - I removed `value` from AudioMuxing and am now relying on the correspondence between the enum names and their WOH values.
  - I removed the `Optional` for audio-muxing, opting for `orElse(NONE)` instead.
  - The code fails faster now. :)

By the way, I'm very open to suggestions regarding the WOH name. I think neither SWITCH nor plapadoo were excited about it. Maybe we use `select-streams` instead?

File: docs/guides/admin/docs/workflowoperationhandlers/select-tracks-woh.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
+SelectTracksWorkflowOperationHandler
```
Comment (KatrinIhler) [inline docs/guides/admin/docs/workflowoperationhandlers/select-tracks-woh.md]: I also think that _SelectStreamsWorkflowOperationHandler_ is clearer.

File: docs/guides/admin/docs/workflowoperationhandlers/select-tracks-woh.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+====================================
+
+
+Description
+-----------
+
+The SelectTracksWorkflowOperationHandler can be used in case not all source tracks should be processed. For example,
+given a recording with a presenter and presentation track, the final recording to be publish may should only include the
```
Comment (KatrinIhler) [inline docs/guides/admin/docs/workflowoperationhandlers/select-tracks-woh.md]: ...and _a_ presentation track, ...to be _published_  _may/should_ ...

File: docs/guides/admin/docs/workflowoperationhandlers/select-tracks-woh.md [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+For example, if there is a media package with two tracks, “presenter” and “presentation”, and the audio stream of
+“presenter” is hidden. Then the WOH will mux presenter’s video stream and presentations audio stream and store it in
+presenter’s place. It will also remove the audio stream from “presentation”.
+
+### `duplicate` ###
+
+The parameter value `duplicate` only applies to media packages that have exactly one non-hidden audio stream. For media
+packages without an audio stream or with more than one audio stream, the behavior is the same as if the parameter were
+omitted. For these media packages, the WOH will mux the audio stream it found to all video streams in the media package.
```
Comment (KatrinIhler) [inline docs/guides/admin/docs/workflowoperationhandlers/select-tracks-woh.md]: Will the audio stream also be muxed into the hidden video streams?

**Review by KatrinIhler — CHANGES_REQUESTED** [review]
I just tested this and found the following things:
1. If I deselect all audio streams, there is an error message, but I can still hit save. and it's computed in the backend. The save button should be disabled.
2. When I deselect all video streams, there's also an error message and I can also still hit save, but then another error message appears. Here the save button should also be disabled.
3. Additionally, the error _You have selected no video tracks to process_ doesn't disappear even if I open a different video in the video editor.

**Comment by pmiddend** [comment]
@KatrinIhler Thanks for the thorough review! I've fixed all of the issues you described.

File: docs/guides/admin/docs/workflowoperationhandlers/index.md [hunk 11]
```diff
@@ -77,6 +77,7 @@ The following table contains the workflow operations that are available in an ou
 |retract-youtube     |Retracts media from YouTube                                    |[Documentation](retract-youtube-woh.md)|
 |segment-video       |Extracting segments from presentation                          |[Documentation](segmentvideo-woh.md)|
 |segmentpreviews     |Extract segment images from a video using FFmpeg               |[Documentation](segmentpreviews-woh.md)|
+|select-tracks       |Select tracks for further processing                           |[Documentation](select-tracks-woh.md)|
```
Comment (gregorydlogan) [inline docs/guides/admin/docs/workflowoperationhandlers/index.md]: select-*tracks*-woh vs select-*streams*-woh.  Should match the rest of the documentation as well (which is probably using both :D)

File: docs/guides/admin/docs/workflowoperationhandlers/select-streams-woh.md [hunk 12]
```diff
@@ -0,0 +1,80 @@
+SelectStreamsWorkflowOperationHandler
+====================================
+
+
+Description
+-----------
+
+The SelectStreamsWorkflowOperationHandler can be used in case not all source tracks should be processed. For example,
```
Comment (gregorydlogan) [inline docs/guides/admin/docs/workflowoperationhandlers/select-streams-woh.md]: track vs stream in this paragraph as well.

File: docs/guides/admin/docs/workflowoperationhandlers/select-streams-woh.md [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+
+Audio Muxing
+-----------------
+
+The optional `audio-muxing` parameter has three possible values: `none` (same as omitting the option), `force` and
+`duplicate`.
+
+### `none` ###
+
+If `none` is specified or the option is omitted, tracks are taken from the specified `source-flavor` and are edited
```
Comment (gregorydlogan) [inline docs/guides/admin/docs/workflowoperationhandlers/select-streams-woh.md]: *audio* tracks I'm assuming?  Maybe rephrase this as "the audio track is taken..."?

File: docs/guides/admin/docs/workflowoperationhandlers/select-streams-woh.md [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+
+\* mandatory configuration key
+
+Audio Muxing
+-----------------
+
+The optional `audio-muxing` parameter has three possible values: `none` (same as omitting the option), `force` and
+`duplicate`.
+
+### `none` ###
```
Comment (gregorydlogan) [inline docs/guides/admin/docs/workflowoperationhandlers/select-streams-woh.md:35]: I'm not a huge fan of this formatting when rendered, but I don't have a better suggestion.  Everything I can find makes things either looks weird or like it's a separate section entirely.

**Comment by gregorydlogan** [comment]
Still working on my review, but I ran into this in testing

```
2018-09-13 11:49:48,360 | ERROR | (WorkflowOperationWorker:180) - Workflow operation 'operation:'image', position:11, state:'FAILED'' failed                                                                                                                                              
java.lang.RuntimeException: Parse failure                                                                                                                                                                                                                                                 
        at com.entwinemedia.fn.parser.Result.getRest(Result.java:50)[25:functional:1.4.2]                                                                                                                                                                                                 
        at org.opencastproject.workflow.handler.composer.ImageWorkflowOperationHandler.parsePositions(ImageWorkflowOperationHandler.java:595)[63:opencast-composer-workflowoperation:6.0.0.SNAPSHOT]                                                                                      
        at org.opencastproject.workflow.handler.composer.ImageWorkflowOperationHandler.configure(ImageWorkflowOperationHandler.java:512)[63:opencast-composer-workflowoperation:6.0.0.SNAPSHOT]                                                                                           
        at org.opencastproject.workflow.handler.composer.ImageWorkflowOperationHandler.start(ImageWorkflowOperationHandler.java:165)[63:opencast-composer-workflowoperation:6.0.0.SNAPSHOT]                                                                                               
        at org.opencastproject.workflow.impl.WorkflowOperationWorker.start(WorkflowOperationWorker.java:233)[193:opencast-workflow-service-impl:6.0.0.SNAPSHOT]                                                                                                                           
        at org.opencastproject.workflow.impl.WorkflowOperationWorker.execute(WorkflowOperationWorker.java:155)[193:opencast-workflow-service-impl:6.0.0.SNAPSHOT]                                                                                                                         
        at org.opencastproject.workflow.impl.WorkflowServiceImpl.runWorkflowOperation(WorkflowServiceImpl.java:892)[193:opencast-workflow-service-impl:6.0.0.SNAPSHOT]                                                                                                                    
        at org.opencastproject.workflow.impl.WorkflowServiceImpl.process(WorkflowServiceImpl.java:1918)[193:opencast-workflow-service-impl:6.0.0.SNAPSHOT]                                                                                                                                
        at org.opencastproject.workflow.impl.WorkflowServiceImpl$JobRunner.call(WorkflowServiceImpl.java:2348)[193:opencast-workflow-service-impl:6.0.0.SNAPSHOT]                                                                                                                         
        at org.opencastproject.workflow.impl.WorkflowServiceImpl$JobRunner.call(WorkflowServiceImpl.java:2314)[193:opencast-workflow-service-impl:6.0.0.SNAPSHOT]                                                                                                                         
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)[:1.8.0_181]                                                                                                                                                                                                           
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)[:1.8.0_181]                                                                                                                                                                                    
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)[:1.8.0_181]                                                                                                                                                                                    
        at java.lang.Thread.run(Thread.java:748)[:1.8.0_181]                                    
```

All I did was ingest a presenter video+audio file, a presentation video file, and then attempt to process to a single presentation+audio output.  I'm using ffmpeg 3.4-static.

File: docs/guides/admin/docs/workflowoperationhandlers/select-streams-woh.md [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+
+If `none` is specified or the option is omitted, tracks are taken from the specified `source-flavor` and are edited
+according to the selections in video editor’s “Tracks” panel. The resulting tracks are stored in the corresponding `
+target-flavor` and `target-tags` are applied.
+
+Editing in the “Tracks” panel usually means removing video or audio streams from tracks. However, if there are multiple
```
Comment (gregorydlogan) [inline docs/guides/admin/docs/workflowoperationhandlers/select-streams-woh.md]: I get what you're trying to say, but this is confusing.  What about something like:

Note: If your editing results in a single video and single audio (track/stream) they will be muxed together even if this option is set to `none`.

File: docs/guides/admin/docs/workflowoperationhandlers/select-streams-woh.md [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+The parameter value `force` only applies to media packages that have exactly one non-hidden audio stream. For media
+packages without an audio stream or with more than one audio stream, the behavior is the same as if the parameter were
+omitted. The same applies to media packages for which there is only one audio stream, and it already belongs to the
+track with flavor type given by `force-target` (or `presenter` if that parameter is omitted).
+
+If, however, there is only one non-hidden audio stream and it does *not* belong to the track given by `force-target`,
+then the WOH will “move” the audio stream to this target track. Specifically, it will mux the video stream of
+`force-target` with the audio stream it found. Then, it removes the audio stream from the original track.
+
+For example, if there is a media package with two tracks, “presenter” and “presentation”, and the audio stream of
```
Comment (gregorydlogan) [inline docs/guides/admin/docs/workflowoperationhandlers/select-streams-woh.md]: For example, consider a mediapackage with two tracks, __presenter__ and __presentation__.  Both of these tracks have audio components, however the __presenter__ audio stream is hidden.  This WOH will mux __presentations__'s audio stream with __presenter__'s video, and remove the audio track from __presentation__'s video. ?

The quotes are used inconsistently, and adding both single and double quotes sucks.  Also reworded slightly.

File: docs/guides/admin/docs/workflowoperationhandlers/select-streams-woh.md [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+“presenter” is hidden. Then the WOH will mux presenter’s video stream and presentations audio stream and store it in
+presenter’s place. It will also remove the audio stream from “presentation”.
+
+### `duplicate` ###
+
+The parameter value `duplicate` only applies to media packages that have exactly one non-hidden audio stream and no
```
Comment (gregorydlogan) [inline docs/guides/admin/docs/workflowoperationhandlers/select-streams-woh.md]: The parameter value `duplicate` only applies to media packages that have exactly one non-hidden audio stream and no hidden video streams.  For these media packages, the WOH will mux the audio stream it found to all video streams in the media package.  For media packages without an audio stream or with more than one audio stream, the behavior is the same as if the parameter were omitted.

Reworded for clarity.

File: docs/guides/admin/docs/workflowoperationhandlers/select-streams-woh.md [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+hidden video streams. For media packages without an audio stream or with more than one audio stream, the behavior
```
Comment (gregorydlogan) [inline docs/guides/admin/docs/workflowoperationhandlers/select-streams-woh.md]: Why does this only work for MPs with non-hidden video streams?  If it's just duplicating from source to N outputs, why can't it handle N-1 outputs?

File: docs/guides/admin/mkdocs.yml [hunk 19]
```diff
@@ -143,6 +143,7 @@ pages:
    - Retract YouTube: 'workflowoperationhandlers/retract-youtube-woh.md'
    - Segment Previews: 'workflowoperationhandlers/segmentpreviews-woh.md'
    - Segment Video: 'workflowoperationhandlers/segmentvideo-woh.md'
+   - Select Tracks: 'workflowoperationhandlers/select-tracks-woh.md'
```
Comment (gregorydlogan) [inline docs/guides/admin/mkdocs.yml]: Tracks vs streams again :)

File: docs/guides/admin/docs/workflowoperationhandlers/select-streams-woh.md [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+-----------------
+
+    <operation
+      id="select-tracks"
+      fail-on-error="true"
+      exception-handler-workflow="partial-error"
+      description="Select tracks for further processing">
+      <configurations>
+        <configuration key="source-flavor">*/source</configuration>
+        <configuration key="target-flavor">*/source</configuration>
```
Comment (gregorydlogan) [inline docs/guides/admin/docs/workflowoperationhandlers/select-streams-woh.md]: Do we want to target source flavours here?  Or something more intermediate?

File: modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/ToolsEndpoint.java [hunk 21]
```diff
@@ -437,26 +441,65 @@ public Response getVideoEditor(@PathParam("mediapackageid") final String mediaPa
...
-            return obj(f("flavor", obj(f("type" ,e.getType()), f("subtype", e.getSubtype()))), f("side", side));
-          })
-          .collect(Collectors.toList());
+    final Map<String, String> latestWfProperties = WorkflowPropertiesUtil
+      .getLatestWorkflowProperties(assetManager, mediaPackageId);
+    // The properties have the format "hide_flavor_audio" or "hide_flavor_video", where flavor is preconfigured.
+    // We filter all the properties that have this format, and then those which have values "true".
+    final Collection<Tuple<String, String>> hiddens = latestWfProperties.entrySet()
+      .stream()
+      .map(p -> Tuple.tuple(p.getKey().split("_"), p.getValue()))
```
Comment (gregorydlogan) [inline modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/ToolsEndpoint.java:447]: Wouldn't it make more sense to do a string comparison for the `hide_flavor` prefix?

File: modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/SelectStreamsWorkflowOperationHandler.java [hunk 22]
```diff
@@ -0,0 +1,529 @@
...
+   *          an instance of the workspace
+   */
+  public void setWorkspace(final Workspace workspace) {
+    this.workspace = workspace;
+  }
+
+  private EncodingProfile getProfile(final String identifier) {
+    final EncodingProfile profile = this.composerService.getProfile(identifier);
+    if (profile == null) {
+      throw new IllegalStateException(String.format("couldn't find encoding profile \"%s\"", identifier));
```
Comment (gregorydlogan) [inline modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/SelectStreamsWorkflowOperationHandler.java]: IIRC there's an exception subclass floating around somewhere for this.

**Review by gregorydlogan — CHANGES_REQUESTED** [review]
Looks good, minor nitpicks.

File: modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/SelectStreamsWorkflowOperationHandler.java (hunk 22 above)
Comment (pmiddend) [reply modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/SelectStreamsWorkflowOperationHandler.java]: Ah yes, the other WOH use `WorkflowOperationException`, so I'll adapt this.

File: modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/ToolsEndpoint.java (hunk 21 above)
Comment (pmiddend) [reply modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/ToolsEndpoint.java:447]: Well, it'll be a little shorter, but I don't think it makes _more_ sense. You'd still have to split the string to get `audio`/`video` and stuff. Or hard-code all the tests altogether.

File: docs/guides/admin/docs/workflowoperationhandlers/select-streams-woh.md (hunk 18 above)
Comment (pmiddend) [reply docs/guides/admin/docs/workflowoperationhandlers/select-streams-woh.md]: Honest reason: Because the code is really complicated as it is. ;) And it's not really properly prepared for videos with more than two tracks (yeah, "tracks" is correct this time :D), which I have stated in the long commentary before the if/else barrage. The plan was to improve on this WOH if somebody has N>2 as a requirement.
If you think it's wrong to merge it in this state, I'll of course rewrite it.

**Comment by pmiddend** [comment]
@gregorydlogan Thanks for the input on the documentation. Incredible how many mistakes I've missed! I've corrected those and used your wording.

File: docs/guides/admin/docs/workflowoperationhandlers/select-streams-woh.md (hunk 18 above)
Comment (staubesv) [reply docs/guides/admin/docs/workflowoperationhandlers/select-streams-woh.md]: @pmiddend & @gregorydlogan Please note that we do explicitly not have the requirement that this code can deal with more than two tracks and it was never our plan to add support for N tracks. The feature is, however, implemented in a way that would allow other Adopters to continue to work on this in case they would want to implement support for N tracks.

File: modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/ToolsEndpoint.java (hunk 21 above)
Comment (gregorydlogan) [reply modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/ToolsEndpoint.java:447]: Ok, fair enough.  On first read I had some minor issue figuring out both the syntax and the goal.  Now that I get the syntax it's easier to read :)

**Comment by gregorydlogan** [comment]
Hi @pmiddend, changes look good.  Still blocked on the processing failure.  Can reproduce on all my development machines.

**Comment by pmiddend** [comment]
@gregorydlogan Okay, I reproduced and fixed it. 

I also fixed the error message you got. Turns out that during exception handling, an exception occured so the string that couldn't be parsed wasn't printed. :D
